components/raftstore/src/store/worker/raftlog_fetch.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use engine_traits::{KvEngine, RaftEngine};
+use raft::GetEntriesContext;
+use std::fmt;
+use tikv_util::worker::Runnable;
+
+use crate::store::{RaftlogFetchResult, SignificantMsg, SignificantRouter};
+
+pub enum Task {
+    PeerStorage {
+        region_id: u64,
+        context: GetEntriesContext,
+        low: u64,
+        high: u64,
+        max_size: usize,
+        tried_cnt: usize,
+        term: u64,
+    },
+    // More to support, suck as fetch entries ayschronously when apply and schedule merge
+}
+
+impl fmt::Display for Task {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Task::PeerStorage {
+                region_id,
+                context,
+                low,
+                high,
+                max_size,
+                tried_cnt,
+                term,
+            } => write!(
+                f,
+                "Fetch Raft Logs [region: {}, low: {}, high: {}, max_size: {}] for sending with context {:?}, tried: {}, term: {}",
+                region_id, low, high, max_size, context, tried_cnt, term,
+            ),
+        }
+    }
+}
+
+pub struct Runner<EK, ER, R>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+    R: SignificantRouter<EK>,
+{
+    router: R,
+    raft_engine: ER,
+    _phantom: std::marker::PhantomData<EK>,
+}
+
+impl<EK: KvEngine, ER: RaftEngine, R: SignificantRouter<EK>> Runner<EK, ER, R> {
+    pub fn new(router: R, raft_engine: ER) -> Runner<EK, ER, R> {
+        Runner {
+            router,
+            raft_engine,
+            _phantom: std::marker::PhantomData,
+        }
+    }
+}
+
+impl<EK, ER, R> Runnable for Runner<EK, ER, R>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+    R: SignificantRouter<EK>,
+{
+    type Task = Task;
+
+    fn run(&mut self, task: Task) {
+        match task {
+            Task::PeerStorage {
+                region_id,
+                low,
+                high,
+                max_size,
+                context,
+                tried_cnt,
+                term,
+            } => {
+                let mut ents = vec![];
+                let res = self.raft_engine.fetch_entries_to(
+                    region_id,
+                    low,
+                    high,
+                    Some(max_size),
+                    &mut ents,
+                );
+
+                let hit_size_limit = res
```

Comment:
If it's an error, then no need to return the entries as next try has to fail.

Reply:
the error would be `ErrCompacted`, so need to return and trigger raft callback.